otlp_json2/__init__.py

```diff
@@ -0,0 +1,173 @@
+    for span in spans:
+        assert span.resource
+        assert span.instrumentation_scope
+        if span.resource is not last_rs:
+            last_rs = span.resource
+            last_is = None
+            rv["resourceSpans"].append(
+                {
+                    "resource": _resource(span.resource),
+                    "scopeSpans": [],
+                }
+            )
+        if span.instrumentation_scope is not last_is:
+            last_is = span.instrumentation_scope
+            rv["resourceSpans"][-1]["scopeSpans"].append(
+                {
+                    "scope": _scope(span.instrumentation_scope),
+                    "spans": [],
+                }
+            )
+        rv["resourceSpans"][-1]["scopeSpans"][-1]["spans"].append(_span(span))
+    return json.dumps(rv, separators=(",", ":")).encode("utf-8")
+
+
+def _resource(resource: Resource):
+    # TODO: add schema_url once that lands in opentelemetry-sdk
+    return _attributes(resource)
+
+
+def _attributes(
+    thing: Resource | InstrumentationScope | ReadableSpan | Event | Link | LogRecord,
+) -> dict[str, Any]:
+    rv = {"attributes": [], "dropped_attributes_count": 0}
+
+    assert thing.attributes is not None
+    for k, v in thing.attributes.items():
+        try:
+            rv["attributes"].append({"key": k, "value": _value(v)})
+        except ValueError:
+            pass
+
+    rv["dropped_attributes_count"] = len(thing.attributes) - len(rv["attributes"])  # type: ignore
+
+    for k in ("attributes", "dropped_attributes_count"):
+        if not rv[k]:
+            del rv[k]
+
+    return rv
+
+
+def _homogeneous_array(value: list[_LEAF_VALUE]) -> list[_LEAF_VALUE]:
+    # TODO: empty lists are allowed, aren't they?
+    if len(types := {type(v) for v in value}) > 1:
+        raise ValueError(f"Attribute value arrays must be homogeneous, got {types=}")
+    return value
+
+
+def _value(value: _VALUE) -> dict[str, Any]:
+    # Attribute value can be a primitive type, excluging None...
+    # protobuf allows bytes, but I think OTLP spec does not?
+    # protobuf allows k:v pairs, but I think OTLP doesn't.
+    # TODO: read up the spec and validate the allowed type range.
+    for klass, (key, post) in _VALUE_TYPES.items():
+        if isinstance(value, klass):
+            return {key: post(value)}
+
+    raise ValueError(f"Cannot convert attribute of {type(value)=}")
+
+
+def _scope(scope: InstrumentationScope):
+    rv = {
+        "name": scope.name,
+        **_attributes(scope),
+    }
+    if scope.version:
+        rv["version"] = scope.version
+    return rv
+
+
+def _span(span: ReadableSpan):
+    assert span.context
+    rv = {
+        "name": span.name,
+        "kind": span.kind.value or 1,  # unspecified -> internal
+        "traceId": _trace_id(span.context.trace_id),
+        "spanId": _span_id(span.context.span_id),
+        "flags": 0x100 | ([0, 0x200][bool(span.parent and span.parent.is_remote)]),
```

Comment:
This is quite obscure:

```python
[0, 0x200][bool(span.parent and span.parent.is_remote)]
```

Clearer (and a bit shorter):

```python
0x200 if span.parent and span.parent.is_remote else 0
```

Also, what are the magic constants 0x100 and 0x200. Should they be named constants, `_FOO_FLAG` and `_BAR_FLAG`?